## Visual Assessment: Acquaint Yourself
This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below.

### Gather

In [1]:
!git clone https://github.com/CherifArsanious/assessing-cleaning-data.git

Cloning into 'assessing-cleaning-data'...
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 6 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (6/6), done.


In [0]:
import pandas as pd

In [0]:
patients = pd.read_csv('assessing-cleaning-data/patients.csv')
treatments = pd.read_csv('assessing-cleaning-data/treatments.csv')
adverse_reactions = pd.read_csv('assessing-cleaning-data/adverse_reactions.csv')

### Assess
In the cells below, each column of each table in this clinical trial dataset is described. To see the table that goes hand in hand with these descriptions, display each table in its entirety by displaying the pandas DataFrame that it was gathered into. This task is the mechanical part of visual assessment in pandas.

In [4]:
# Display the patients table
patients

patient_id assigned_sex given_name  ... weight height   bmi
0             1       female        Zoe  ...  121.7     66  19.6
1             2       female     Pamela  ...  118.8     66  19.2
2             3         male        Jae  ...  177.8     71  24.8
3             4         male       Liêm  ...  220.9     70  31.7
4             5         male        Tim  ...  192.3     27  26.1
..          ...          ...        ...  ...    ...    ...   ...
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[503 rows x 14 columns]

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [5]:
patients.head()

patient_id assigned_sex given_name  surname  ...  birthdate weight height   bmi
0           1       female        Zoe  Wellish  ...  7/10/1976  121.7     66  19.6
1           2       female     Pamela     Hill  ...   4/3/1967  118.8     66  19.2
2           3         male        Jae   Debord  ...  2/19/1980  177.8     71  24.8
3           4         male       Liêm     Phan  ...  7/26/1951  220.9     70  31.7
4           5         male        Tim  Neudorf  ...  2/18/1928  192.3     27  26.1

[5 rows x 14 columns]

In [6]:
patients.tail()

patient_id assigned_sex given_name  ... weight height   bmi
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[5 rows x 14 columns]

In [7]:
patients.sample()

patient_id assigned_sex given_name surname  ... birthdate weight height   bmi
413         414         male        Kai   Cunha  ...  2/1/1990  163.5     74  21.0

[1 rows x 14 columns]

In [8]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [9]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [10]:
# Display the treatments table
treatments

given_name       surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0     veronika      jindrová  41u - 48u  ...        7.63       7.20           NaN
1       elliot    richardson          -  ...        7.56       7.09          0.97
2     yukitaka      takenaka          -  ...        7.68       7.25           NaN
3         skye    gormanston  33u - 36u  ...        7.97       7.62          0.35
4       alissa        montez          -  ...        7.78       7.46          0.32
..         ...           ...        ...  ...         ...        ...           ...
275     albina      zetticci  45u - 51u  ...        7.93       7.73          0.20
276       john   teichelmann          -  ...        7.90       7.58           NaN
277     mathea       lillebø  23u - 36u  ...        9.04       8.67          0.37
278     vallie        prince  31u - 38u  ...        7.64       7.28          0.36
279     samúel  guðbrandsson  53u - 56u  ...        8.00       7.64          0.36

[280 rows x 7 columns]

In [11]:
patients.query('city=="New York"').shape[0]

18

In [12]:
patients.query('surname=="Doe"').shape[0]

6

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [14]:
# Display the adverse_reactions table
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

In [0]:
adverse_reactions['adverse_reaction'].value_counts()

## Cleaning Data

In [0]:
# Making copies of our dataframes 
patients_clean=patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

## treatments: Missing records (280 instead of 350)
### Define
merge the two tables the one of treatments and the other one treatments_cut that holds the rest of records

In [0]:
df_cut=pd.read_csv('assessing-cleaning-data/treatments_cut.csv')

In [51]:
df_cut.head()

given_name    surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0      jožka  resanovič  22u - 30u  ...        7.56       7.22          0.34
1  inunnguaq   heilmann  57u - 67u  ...        7.85       7.45           NaN
2      alwin   svensson  36u - 39u  ...        7.78       7.34           NaN
3        thể      lương          -  ...        7.64       7.22          0.92
4     amanda    ribeiro  36u - 44u  ...        7.85       7.47          0.38

[5 rows x 7 columns]

In [52]:
df_cut.shape[0]

70

In [53]:
treatments_clean.shape[0]

280

In [0]:
treatments_clean=pd.concat([treatments_clean,df_cut],axis=0,ignore_index=True)

In [55]:
treatments_clean.index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [56]:
treatments_clean.shape[0]

350

### treatments: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one Define, Code, and Test header occurs multiple times in this notebook.

In [57]:
treatments_clean.head()

given_name     surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0   veronika    jindrová  41u - 48u  ...        7.63       7.20           NaN
1     elliot  richardson          -  ...        7.56       7.09          0.97
2   yukitaka    takenaka          -  ...        7.68       7.25           NaN
3       skye  gormanston  33u - 36u  ...        7.97       7.62          0.35
4     alissa      montez          -  ...        7.78       7.46          0.32

[5 rows x 7 columns]

In [0]:
treatments_clean['hba1c_change']=treatments_clean['hba1c_start']-treatments_clean['hba1c_end']

In [59]:
treatments_clean['hba1c_change'].isnull().sum()

0

## Contact column in patients table contains two variables: phone number and email
Extract the phone number and email variables from the contact column using regular expressions and pandas' str.extract method. Drop the contact column when done.

In [0]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

In [61]:
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [62]:
patients_clean.phone_number.sample(25)

482         361-693-4960
336         248-978-1574
320         262-565-9052
328    +1 (612) 342-6065
303         313 589 1746
66          760-287-8295
481         918 706 2776
200         856-655-5415
217         706-266-4583
102    +1 (334) 261-1235
206         806-323-2210
358         203 235 1076
331         323 635 9919
160         602-993-7880
231         252 583 5410
68          414-646-4353
370         508-821-2421
9           718 795 9124
384    +1 (605) 440-5492
238         215-528-2193
151         916-224-7868
473         508 905 2371
110    +1 (214) 637-0855
79          302-397-6035
211         612 228 4170
Name: phone_number, dtype: object

In [63]:
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

## Define
Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.

In [64]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean


given_name     surname  hba1c_start  ...  hba1c_change  treatment       dose
0     veronika    jindrová         7.63  ...          0.43    auralin  41u - 48u
1       elliot  richardson         7.56  ...          0.47    auralin          -
2     yukitaka    takenaka         7.68  ...          0.43    auralin          -
3         skye  gormanston         7.97  ...          0.35    auralin  33u - 36u
4       alissa      montez         7.78  ...          0.32    auralin          -
..         ...         ...          ...  ...           ...        ...        ...
695     rovzan     kishiev         7.75  ...          0.34    novodra          -
696      jakob    jakobsen         7.96  ...          0.45    novodra  28u - 26u
697      bernd   schneider         7.74  ...          0.30    novodra          -
698      berta  napolitani         7.68  ...          0.47    novodra  42u - 44u
699     armina       sauvé         7.86  ...          0.46    novodra          -

[700 rows x 7 columns]

In [65]:
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean.head()

given_name     surname  hba1c_start  ...  hba1c_change  treatment       dose
0   veronika    jindrová         7.63  ...          0.43    auralin  41u - 48u
3       skye  gormanston         7.97  ...          0.35    auralin  33u - 36u
6     sophia      haugen         7.65  ...          0.38    auralin  37u - 42u
7      eddie      archer         7.89  ...          0.34    auralin  31u - 38u
9       asia     woźniak         7.76  ...          0.39    auralin  30u - 36u

[5 rows x 7 columns]

In [66]:
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean=treatments_clean.drop('dose',axis=1)
treatments_clean.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


given_name     surname  hba1c_start  ...  treatment  dose_start dose_end
0   veronika    jindrová         7.63  ...    auralin         41u      48u
3       skye  gormanston         7.97  ...    auralin         33u      36u
6     sophia      haugen         7.65  ...    auralin         37u      42u
7      eddie      archer         7.89  ...    auralin         31u      38u
9       asia     woźniak         7.76  ...    auralin         30u      36u

[5 rows x 8 columns]

## Adverse reaction should be part of the treatments table
### Define
Merge the adverse_reaction column to the treatments table, joining on given name and surname.

In [67]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')
#adding argument how='left' to preserve all the rows of the treatments 
#table (the left table)as not all patients suffered from adverse effects so a normal 
#merge without the how left parameter would have 
#merged only the common rows between the two tables
treatments_clean.head(25)

given_name       surname  ...  dose_end   adverse_reaction
0      veronika      jindrová  ...       48u                NaN
1          skye    gormanston  ...       36u                NaN
2        sophia        haugen  ...       42u                NaN
3         eddie        archer  ...       38u                NaN
4          asia       woźniak  ...       36u                NaN
5        joseph           day  ...       36u       hypoglycemia
6       roxanne     andreyeva  ...       38u                NaN
7        simone  baumgaertner  ...       37u                NaN
8          enco        žibrik  ...       68u                NaN
9       camilla      zaitseva  ...       37u                NaN
10        tekla       walczak  ...       39u                NaN
11  brancaleone         russo  ...       60u                NaN
12         isac          berg  ...       41u                NaN
13      clinton        miller  ...       51u  throat irritation
14       eugene       mironov  ...       49u                NaN
15  szilveszter         totth  ...       39u                NaN
16    alexander     mathiesen  ...       58u                NaN
17          chỉ           lâm  ...       48u                NaN
18    władysław     wieczorek  ...       37u                NaN
19     kristján       ingason  ...       55u                NaN
20       marija      grubišić  ...       43u                NaN
21        sauli    koivuniemi  ...       47u                NaN
22      mariana         souza  ...       42u                NaN
23   kristoffer     martinsen  ...       37u                NaN
24           mỹ         quynh  ...       64u                NaN

[25 rows x 9 columns]

In [68]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:5208: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [70]:
treatments_clean

hba1c_start  hba1c_end  ...           adverse_reaction patient_id
0           7.63       7.20  ...                        NaN        225
1           7.97       7.62  ...                        NaN        242
2           7.65       7.27  ...                        NaN        345
3           7.89       7.55  ...                        NaN        276
4           7.76       7.37  ...                        NaN         15
..           ...        ...  ...                        ...        ...
344         7.51       7.06  ...                     nausea        153
345         7.67       7.30  ...                        NaN        420
346         9.21       8.80  ...  injection site discomfort        336
347         7.96       7.51  ...               hypoglycemia         25
348         7.68       7.21  ...  injection site discomfort        477

[349 rows x 8 columns]